### ISSUE TITLE
Cannot register multiple load hooks

### ISSUE THREAD

**Issue by casparvl** [TACC/Lmod#694]
**The issue**
I'm a collaborator on the [EESSI](https://www.eessi.io/docs/) project, where we try to build a common, shared software stack that can be used on a wide range of systems. We provide an `lmod`-based module environment, and are implementing some of our logic in `lmod` hooks. The details are not that important (though you can see [here](https://github.com/EESSI/software-layer/blob/a28e5ee6ff295cece13c7d632bf24f161b07bfd8/create_lmodrc.py#L32) the exact definition of our `lmodrc.lua` file), but we basically have two functions that we would like to be executed and naively implemented it like this:

```
-- lmodrc.lua

local function eessi_hook_a(t)
    <do_something>
end
local function eessi_hook_b(t)
    <do_something_else>
end
hook.register("load", eessi_hook_a)
hook.register("load", eessi_hook_b)
```

Somewhat to our surprise, the only thing that got execute was `eessi_hook_b`. Looking at the sources, I found [here](https://github.com/TACC/Lmod/blob/a97d68193a6f32ebca4b4bd70dac0f6ac8fddefe/src/Hook.lua#L87) that this is easily understood, since a call to `.register` just overwrites the `load` hook if it already exists.

Of course, for our own hooks, it can easily be solved / worked around by first combining everything in a single function, and then registering that as a load hook:
```
-- eessi lmodrc

local function eessi_hook_a(t)
    <do_something>
end
local function eessi_hook_b(t)
    <do_something_else>
end
function eessi_load_hook(t)
    eessi_hook_a(t)
    eessi_hook_b(t)
end
hook.register("load", eessi_load_hook)
```

The real challenge is that we _also_ want to facilitate sites to add site-specific things (e.g. site-specific configuration through setting `OMPI_MCA` variables whenever an `OpenMPI` module gets loaded). From [these docs](https://lmod.readthedocs.io/en/latest/145_properties.html?highlight=lmod_rc#the-properties-file-lmodrc-lua) it is clear that `lmodrc` files can be merged from different levels, which facilitates exactly this site-specific tuning. We plan to define the default EESSI hooks in a `$LMOD_CONFIG_DIR/lmodrc.lua` file, and instruct sites to set `$LMOD_RC` to provide an `lmodrc.lua` in which they can implement additional hooks. In this scenario, the above solution doesn't work anymore, or at the very lease becomes extremely ugly. To combine things into a single function, we need to import the EESSI `lmodrc` file (from `$LMOD_CONFIG_DIR/lmodrc.lua`) into the site-specific `lmodrc` file (set with `$LMOD_RC`), create a single function again, and register it as the new `load` hook:

```
-- site-specific lmodrc
local function site_specific_hook_a(t)
    <do_something>
end

lmod_config_dir = os.getenv("LMOD_CONFIG_DIR")
dofile(lmod_config_dir.."/lmodrc.lua")

local function combined_load_hook(t)
    eessi_load_hook(t)
    site_specific_hook_a(t)
end
hook.register("load", combined_load_hook)
```

This approach feels quite unnatural and hacky: I need to _know_ the original function name `eessi_load_hook` that I want to include in the `combined_load_hook`, I need to handle cases where `LMOD_CONFIG_DIR` isn't set, or the `lmodrc.lua` file in that dir isn't found... All in all, it feels like the limitation of only being able to register one function as a hook does not work well together with the ability to combine multiple `lmodrc` files in a hierarchical way.

**Potential solution 1**
I think the _easiest_ solution would be to allow multiple hooks to be registered. I'm definitely no LUA expert, but I'm assuming the `hook.register` function could also just append functions to a list. The `hook.apply` would then have to be changed so that it executes all functions in the list (probably in the order in which they were registered). The EESSI and site-specific `lmodrc` files would then essentially have:
```
-- eessi lmodrc
...
hook.register("load", eessi_hook_a)
hook.register("load", eessi_hook_b)
```
```
-- site-specific lmodrc
...
hook.register("load", site_specific_hook_a)
```
Upon loading a module, this would then execute `eessi_hook_a`, `eessi_hook_b` and `site_specific_hook_a` (in that order). Note that in this approach, you'd probably also want some way to remove existing hooks, for scenarios where one _did_ intend to overwrite an old hook. (something like a `hook.remove("load", <index>)`, where <index> is an index to the list of functions)

**Potential solution 2**
Another solution - though I personally like this less - would be to provide a public function that allows one to retreive the current function that is registered as a hook, so that one can append to it. 

```
-- Hook.lua
...
function M.read(name)
    return validT[name]
end
```
```
-- site-specific lmodrc
local original_load_hook = hook.read("load")
local function site_specific_hook(t)
    original_load_hook(t)
    <do_something>
end

hook.register("load", site_specific_hook)
```
Again, I'm no Lua expert, so I'm not 100% sure you can return / assign / use functions in this way in Lua, but I think so.

I'm happy to have a look if I can contribute this, but figured I'd first open the discussion to see a) if you are open to this and b) which of the two approaches you'd consider preferable.

**Comment by casparvl** [comment]
We just had a discussion in the Lmod monthly meeting. Let me try to summarize what we discussed (feel free to correct me if I'm wrong). First, I'll make

```
-- lmodrc.lua

local function eessi_hook_a(t)
    <do_something>
end
local function eessi_hook_b(t)
    <do_something_else>
end
hook.register("load", eessi_hook_a)
hook.register("load", eessi_hook_b)
```
behave as we initially expected, i.e. that it executes both `eessi_hook_a` and `eessi_hook_b`. I'll do this by making sure that internally, these get appended to a list (and `hook.apply` will need to then loop over all of those). Next step would then be to make sure that the behaviour depends on the 2nd argument type. If that is a function, we keep the old behaviour (overwrite the current hook, essentially as a list of length 1). Then, extend the API so that it can also take `hook.register("load", {some_table})`. If it gets called through this interface, we append instead of overwrite.

**Comment by rtmclay** [comment]
This issue has been solved and is now part of Lmod 8.7.36
